File: UICSpring17/CS412/hw1programming/mnist1NNdemo.m
%% Classify the MNIST digits using a one nearest neighbour classifier and Euclidean distance
%% This file is modified from pmtk3.googlecode.com

load('mnistData');
clear
% set training & testing 
sizes = [100, 200, 500, 1000, 2000, 5000, 10000];
accuracy = [];
foundNonMatch = 0;


for i = sizes
  trainndx = 1:i; 
  testndx =  1:i; 
  ntrain = length(trainndx);
  ntest = length(testndx);
  Xtrain = double(reshape(mnist.train_images(:,:,trainndx),28*28,ntrain)');
  Xtest  = double(reshape(mnist.test_images(:,:,testndx),28*28,ntest)');

  ytrain = (mnist.train_labels(trainndx));
  ytest  = (mnist.test_labels(testndx));

  % Precompute sum of squares term for speed
  XtrainSOS = sum(Xtrain.^2,2);
  XtestSOS  = sum(Xtest.^2,2);

  % fully solution takes too much memory so we will classify in batches
  % nbatches must be an even divisor of ntest, increase if you run out of memory 
  if ntest > 1000
  nbatches = 50;
  else
  nbatches = 5;
  end
  batches = mat2cell(1:ntest,1,(ntest/nbatches)*ones(1,nbatches));
  ypred = zeros(ntest,1);
  % Classify
  for i=1:nbatches    
  dst = sqDistance(Xtest(batches{i},:),Xtrain,XtestSOS(batches{i},:),XtrainSOS);
  [junk,closest] = min(dst,[],2);
  ypred(batches{i}) = ytrain(closest);
  end
  
  errorRate = mean(ypred ~= ytest);
  fprintf('Error Rate: %.2f%%\n',100*errorRate);
  accuracy = [accuracy errorRate];
 endfor
  % Report


%%% Plot example
figure();
% line plot example random data
plot( sizes,accuracy);
ylabel('accuracy');
xlabel('size');

%figure();
% image plot
%imshow(mnist.train_images(:,:,3)); % plot the third image

for k = 1:length(ypred)
  if ypred(k) != ytest(k) && foundNonMatch == 0
    foundNonMatch = 1;
    figure();
    imshow(mnist.train_images(:,:,k));
    figure();
    imshow(mnist.test_images(:,:,k));
  end
end

File: UICSpring17/CS412/hw1programming/mnist1NNdemo.jl
using MAT

# function for calculating squared distance
function sqdistance(p, q, pSOS, qSOS)
  return broadcast(+, pSOS, qSOS') - 2*p*q'
end

# open file
file = matopen("mnistData.mat")
mnist = read(file, "mnist")

# load data
train_images = mnist["train_images"]
train_labels = mnist["train_labels"]
test_images = mnist["test_images"]
test_labels = mnist["test_labels"]

# set training & testing
trainndx = 1:10000
testndx =  1:10000

ntrain = length(trainndx)
ntest = length(testndx)
Xtrain = float(reshape(train_images[:,:,trainndx], 28*28, ntrain)')
Xtest  = float(reshape(test_images[:,:,testndx], 28*28, ntest)')

ytrain = train_labels[trainndx]
ytest  = test_labels[testndx]

# Precompute sum of squares term
XtrainSOS = sum(Xtrain.^2, 2)
XtestSOS  = sum(Xtest.^2, 2)

# fully solution takes too much memory so we will classify in batches
# nbatches must be an even divisor of ntest, increase if you run out of memory
if ntest > 1000
  nbatches = 50
  else
  nbatches = 5
end

nel = ntest ÷ nbatches
batches = [nel*(i-1)+1:nel*i for i = 1:nbatches]

ypred = zeros(UInt8, ntest)
# classify
for i=1:nbatches
  dist = sqdistance(Xtest[batches[i],:], Xtrain, XtestSOS[batches[i]], XtrainSOS)
  for j = 1:length(batches[i])
    closest = indmin(dist[j,:])
    ypred[batches[i][j]] = ytrain[closest]
  end
end

error_rate = mean(ypred .!= ytest)

println("Error Rate : ", 100 * error_rate)


### Plot example
using Plots
pyplot()

# line plot example random data
plot(10*rand(10,1), label="accuracy")

# image plot
img = flipdim(train_images[:,:,3], 1)  # plot the third image
heatmap(1:28, 1:28, img, fillcolor = :grays, legend = false)
Plots.gui()
